### PR TITLE
Fix `AggregateStatistics` optimization so it doesn't change output type

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/2673

 # Rationale for this change
See https://github.com/apache/arrow-datafusion/issues/2673 -- the optimizer pass is changing input types.

# What changes are included in this PR?
1. Fix bug
2. New Regresion test
3. Give some constants related to `COUNT(*)` expansion symbolic names to improve readability

# Are there any user-facing changes?
less bugs

# Does this PR break compatibility with Ballista?
No

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/aggregate_statistics.rs [hunk 1]
```diff
@@ -148,10 +152,10 @@ fn take_optimizable_table_count(
                 .as_any()
                 .downcast_ref::<expressions::Literal>()
             {
-                if lit_expr.value() == &ScalarValue::UInt8(Some(1)) {
+                if lit_expr.value() == &COUNT_STAR_EXPANSION {
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/aggregate_statistics.rs:155]: There was an implicit coupling between the SQL planner and this file, which I have now made explicit with a named constant

File: datafusion/core/src/physical_optimizer/aggregate_statistics.rs [hunk 2]
```diff
@@ -293,38 +297,80 @@ mod tests {
...
-                .downcast_ref::<UInt64Array>()
+                .downcast_ref::<Int64Array>()
                 .unwrap()
                 .values(),
             &[count]
         );
+
+        // Validate that the optimized plan returns the exact same
+        // answer (both schema and data) as the original plan
+        let task_ctx = session_ctx.task_ctx();
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/aggregate_statistics.rs:326]: This test would have caught this issue when it was introduced in https://github.com/apache/arrow-datafusion/pull/2636

File: datafusion/core/src/physical_optimizer/aggregate_statistics.rs [hunk 3]
```diff
@@ -293,38 +297,80 @@ mod tests {
...
+        let plan_result = common::collect(plan.execute(0, task_ctx)?).await?;
+        assert_eq!(normalize(result), normalize(plan_result));
         Ok(())
     }
 
+    /// Normalize record batches for comparison:
+    /// 1. Sets nullable to `true`
+    fn normalize(batches: Vec<RecordBatch>) -> Vec<RecordBatch> {
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/aggregate_statistics.rs:334]: This is stupid but necessary to pass the tests

File: datafusion/core/src/physical_optimizer/aggregate_statistics.rs [hunk 4]
```diff
@@ -293,38 +297,80 @@ mod tests {
...
+                COUNT_STAR_NAME.to_string(),
+            ),
+            Some(s) => (
+                expressions::col(col.unwrap(), s).unwrap(),
+                format!("COUNT({})", col.unwrap()),
+            ),
         };
-        Arc::new(Count::new(expr, "my_count_alias", DataType::UInt64))
+
+        Arc::new(Count::new(expr, name, DataType::Int64))
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/aggregate_statistics.rs:373]: Now that the schema is checked, we can't use some arbitrary column name, we need to use the actual name the plan would

File: datafusion/core/src/physical_optimizer/aggregate_statistics.rs [hunk 5]
```diff
@@ -148,10 +152,10 @@ fn take_optimizable_table_count(
                 .as_any()
                 .downcast_ref::<expressions::Literal>()
             {
-                if lit_expr.value() == &ScalarValue::UInt8(Some(1)) {
+                if lit_expr.value() == &COUNT_STAR_EXPANSION {
                     return Some((
-                        ScalarValue::UInt64(Some(num_rows as u64)),
-                        "COUNT(UInt8(1))",
+                        ScalarValue::Int64(Some(num_rows as i64)),
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/aggregate_statistics.rs:157]: The change from `UInt64` to `Int64` here and a few lines below is the actual bug fix / change of behavior -- the rest of this PR is testing / readability improvements

File: datafusion/core/src/physical_optimizer/aggregate_statistics.rs [hunk 6]
```diff
@@ -293,38 +297,80 @@ mod tests {
     /// Checks that the count optimization was applied and we still get the right result
     async fn assert_count_optim_success(plan: AggregateExec, nulls: bool) -> Result<()> {
         let session_ctx = SessionContext::new();
-        let task_ctx = session_ctx.task_ctx();
         let conf = session_ctx.copied_config();
-        let optimized = AggregateStatistics::new().optimize(Arc::new(plan), &conf)?;
+        let plan = Arc::new(plan) as _;
+        let optimized = AggregateStatistics::new().optimize(Arc::clone(&plan), &conf)?;
 
         let (col, count) = match nulls {
```
Comment (tustvold) [inline datafusion/core/src/physical_optimizer/aggregate_statistics.rs:304]: I was very confused by what this parameter controls, should it not be something like `column: Option<&str>` instead?

File: datafusion/core/src/physical_optimizer/aggregate_statistics.rs [hunk 7]
```diff
@@ -293,38 +297,80 @@ mod tests {
...
                 .unwrap()
                 .values(),
             &[count]
         );
+
+        // Validate that the optimized plan returns the exact same
+        // answer (both schema and data) as the original plan
+        let task_ctx = session_ctx.task_ctx();
+        let plan_result = common::collect(plan.execute(0, task_ctx)?).await?;
+        assert_eq!(normalize(result), normalize(plan_result));
```
Comment (tustvold) [inline datafusion/core/src/physical_optimizer/aggregate_statistics.rs:328]: A couple of lines above there is

```
assert_eq!(result[0].schema(), Arc::new(Schema::new(vec![col])));
```

This would suggest to me that the result has a single column and a single field. Perhaps we could just do something like

```
let expected_a_schema = ..;
let expected_b_schema = ..;
for (a, b) in result.iter().zip(plan_result) {
  assert_eq!(a.column(0), b.column(0);
  assert_eq!(a.schema(), expected_a_schema);
  assert_eq!(b.schema(), expected_b_schema);
}
```

I think the normalization logic is a little bit hard to follow...

**Review by tustvold — APPROVED** [review]
Looks good to me, minor comment about test readability, and a comment about nullability.

I also wonder if the fact it isn't always not nullable is actually a bug in the aggregate function, can `count(..)` ever return NULL? I thought it just skipped nulls?
